FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1069 @@
+#include "FEXCore/Utils/LogManager.h"
+#include "Interface/Core/Interpreter/Fallbacks/FallbackOpHandler.h"
+#include "Interface/IR/IR.h"
+#include "Interface/IR/IREmitter.h"
+#include "Interface/IR/PassManager.h"
+#include <FEXCore/IR/IR.h>
+#include <FEXCore/Utils/Profiler.h>
+#include <FEXCore/fextl/deque.h>
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <stdint.h>
+
+// This file adds a pass to process X87 stack instructions.
+// These instructions are marked in IR.json with `X87: true` and are generated
+// by X87 guest instructions.
+// The way is works is that there's a virtual stack `StackData`, where we load and store
+// and apply the operations in a block of code. Once the block finishes, we emit the necessary operations
+// that we recorded onto the virtual stack. This allows us to save a lot of code movement
+// to and from stack registers, top management and valid flags. It also allows us to
+// perform memcpy optimizations like the one performed in STORESTACKMEMORY.
+//
+// By default we run on the fast path - i.e. we assume all values are in the stack and we have a complete
+// stack overview. However, if we encounter a value that's not in the virtual stack - maybe it was added
+// to the stack in a previous block, we move onto the slow path which loads and stores values to the stack
+// registers.
+// Once in a slow path, we won't return to the fast pass until the beginning of the following block.
+
+namespace FEXCore::IR {
+
+// FIXME(pmatos): copy from OpcodeDispatcher.h
+inline uint32_t MMBaseOffset() {
+  return static_cast<uint32_t>(offsetof(Core::CPUState, mm[0][0]));
+}
+
+// Similar helper to the one in OpcodeDispatcher.h except we do not
+// need to handle flags, etc.
+template<typename T>
+void DeriveOp(Ref& RefV, IROps NewOp, IREmitter::IRPair<T> Expr) {
+  Expr.first->Header.Op = NewOp;
+  RefV = Expr;
+}
+
+enum class StackSlot { UNUSED, INVALID, VALID };
+// FixedSizeStack is a model of the x87 Stack where each element in this
+// fixed size stack lives at an offset from top. The top of the stack is at
+// index 0.
+template<typename T>
+class FixedSizeStack {
+public:
+  static constexpr uint8_t size = 8;
+
+  // Real top as an offset from stored top value (or the one at the beginning of the block)
+  // For example, if we start and push a value to our simulated stack, because we don't
+  // update top straight away the TopOffset is 1.
+  // If SlowPath is true, then TopOffset is always zero.
+  int8_t TopOffset = 0;
+
+  FixedSizeStack()
+    : buffer(FixedSizeStack::size, {StackSlot::UNUSED, T()}) {}
+
+  void push(const T& Value) {
+    rotate();
+    buffer.front() = {StackSlot::VALID, Value};
```

Comment:
Now that we're using `vector`, we can invert the storage order of the elements to simplify the implementation. Instead of `rotate` + `assign to front()`, we can just use this:
```suggestion
    buffer.push_back({StackSlot::VALID, Value});
```
Similarly, pop becomes `pop_back()`, and `StackSlot::UNUSED` can be entirely removed.

Internally, the buffer won't be fixed-size anymore then, but that's fine. Calling `reserve(size)` on construction ensures there will never be more than one heap allocation.

Reply:
I understand the suggestion but the location where the items are valid correspond to an offset to TOP, which will stop being the case. I understand that this optimization is possible, but I think it's going to complicate the code readability elsewhere. I will take a look to see exactly what such a change would look like and report back. Thanks.

Reply:
Another thing is that we cannot really rmeove StackSlot::UNUSED. This is because the stack doesn't need to be contiguous, i.e. some elements might be unused. So you do need to mark that. For example:
```
fld qword [addr] ; 1.0
fst st2
```

In this case the stack will look like:
```
MM1  1.0      ST2
MM0  UNUSED   ST1
MM7  1.0      ST0 <- TOP
```

So we need to mark these in-between slots as unused.

Reply:
Are we baking pointers within the FixedSizeStack into the generated assembly code? Otherwise I don't understand why you're bringing up assembly code here.

The change I'm proposing is an implementation detail and shouldn't affect any of the users of the FixedSizeStack interface.

Reply:
Maybe I am confused about your suggestion, but I cannot see how your suggestion would work. The "stack" is not really a stack, which makes it harder to implement as you suggest. This is because I can push an element to the stack but then set the element that's two elements above that. With your proposed solution where the vector starts empty a push would be an emplace_back but then setting ST2, would mean a rotate right by one and emplacing back. It's certainly possible but I am not sure it's an improvement.

Reply:
> The "stack" is not really a stack

I think that's what confused me here. I see now that with the offset-parameter in top/setTop, we really always need 8 valid elements (no more, no less), which means you can't just push elements without also dropping the last element.

Bummer :(